exp/controllers/machinepool_controller.go

```diff
@@ -137,7 +137,12 @@ func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rete
 		// TODO(jpang): add support for metrics.
 
 		// Always attempt to patch the object and status after each reconciliation.
-		if err := patchHelper.Patch(ctx, mp); err != nil {
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
```

Comment:
It looks like there is at least one case (possibly multiple) where this would cause setting the observed generation prior to actually reconciling the resource fully.

The first instance I was able to find is where we are adding the finalizer to the resource and returning early

Reply:
xposting here becuase the topic is the same https://github.com/kubernetes-sigs/cluster-api/pull/3226#discussion_r446940535